outlines/generate/generator.py

```diff
@@ -0,0 +1,315 @@
+import dataclasses
+import math
+from typing import TYPE_CHECKING, Callable, List, Optional, Union
+
+import torch
+
+from outlines.index.index import FSMState
+
+if TYPE_CHECKING:
+    from outlines.generate.samplers import Sampler
+    from outlines.index.index import FSM
+
+
+@dataclasses.dataclass(frozen=True)
+class GenerationState:
+    token_ids: torch.Tensor
```

Comment:
Depending on how expensive `concatenate` is it might make sense to carry lists of `Sequence` objects so we can handle beam search & SMC more easily.